mindmaps-migration/json/src/main/java/io/mindmaps/migration/json/JsonSchemaMigrator.java

```diff
@@ -0,0 +1,245 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with MindmapsDB. If not, see <http://www.gnu.org/licenses/gpl.txt>.
+ */
+
+package io.mindmaps.migration.json;
+
+
+import io.mindmaps.MindmapsGraph;
+import io.mindmaps.concept.EntityType;
+import io.mindmaps.concept.ResourceType;
+import io.mindmaps.concept.RoleType;
+import io.mindmaps.concept.Type;
+import mjson.Json;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Stream;
+
+import static io.mindmaps.concept.ResourceType.DataType.STRING;
+import static java.util.stream.Collectors.joining;
+import static java.util.stream.Collectors.toList;
+import static java.util.stream.Collectors.toSet;
+
+/**
+ * Migrator for migrating a JSON schema into a Mindmaps ontology
+ */
+public class JsonSchemaMigrator {
+
+    private MindmapsGraph graph;
+    private static final String ID_SUFFIX = ".json#";
+    private Map<String, Type> migrated;
+
+    // WARNING: This regex cannot perfectly match all emails because the definition is context-free
+    private static final String EMAIL_REGEX =
+            "([!#-'*+/-9=?A-Z^-~-]+(\\.[!#-'*+/-9=?A-Z^-~-]+)*|\"([]!#-[^-~ \t]|(\\[\t -~]))+\")@" +
+                    "([!#-'*+/-9=?A-Z^-~-]+(\\.[!#-'*+/-9=?A-Z^-~-]+)*|\\[[\t -Z^-~]*])";
+
+    /**
+     * Create a JsonSchemaMigrator to migrate into the given graph
+     */
+    public JsonSchemaMigrator() {
+        migrated = new HashMap<>();
+    }
+
+    public JsonSchemaMigrator graph(MindmapsGraph graph){
+        this.graph = graph;
+        return this;
+    }
+
+    /**
+     * Migrate a JSON schema into a mindmaps ontology
+     * @param schema a JSON schema
+     * @return the root concept type representing the schema
+     */
+    public Type migrateSchema(Json.Schema schema) {
+        String id = schema.toJson().at("id").asString();
+        if (id.startsWith("/")) id = id.substring(1);
+        String name = id.substring(0, id.length() - ID_SUFFIX.length()).replace("/", "-");
+        return migrateSchema(name, schema);
+    }
+
+    /**
+     * Migrate a JSON schema into a mindmaps ontology
+     * @param name the name of the schema
+     * @param schema a JSON schema
+     * @return the root concept type representing the schema
+     */
+    public Type migrateSchema(String name, Json.Schema schema) {
+        return migrateSchema(name, schema.toJson());
+    }
+
+    private Type migrateSchema(String name, Json schema) {
+        if(migrated.containsKey(name)){
+            return migrated.get(name);
+        }
+        
+        Set<JsonType> types = getTypes(schema).collect(toSet());
+
+        // Check if any subtypes are resource types
+        boolean isResource = types.stream().anyMatch(JsonType::isResourceType);
```

Comment:
Move this further down after your initial ifs just before isResource is needed. That way you prevent yourself from iterating over this needlessly.